Added support for RGB24 (RGB3)

GilesBathgate opened the pull request:
This adds support for RGB24 aka RGB3. It's only needed by some very rare cameras.

Fixes #248

GilesBathgate committed:
Added support for RGB24 (RGB3)

jacksonliam commented:
This is great, nice one!

Just one bit missing - could you add RGB24 to the -fourcc command line option too? So it can be forced and doesn't need to "fall back".

https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/plugins/input_uvc/input_uvc.c#L319

Make sure to update the help text

https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/plugins/input_uvc/input_uvc.c#L543

Cheers!

GilesBathgate committed:
Added -fourcc RGB24 option and updated documentation

GilesBathgate commented:
I've added the -fourcc option and updated the help text.

Here is a screenshot:
![image](https://user-images.githubusercontent.com/444143/80283417-4760b200-870f-11ea-88e7-4708275973ee.png)

jacksonliam commented:
Thanks very much for this, I was able to verify it with a Raspberry pi camera and ` bcm2835-v4l2` as it supports RGB3.

The picture passed a brief quality/colour test, so I believe the conversion is done correctly. I believe it is a tiny bit darker than the YUV mode captures I compared it with but this might be something to do with ` bcm2835-v4l2`.

GilesBathgate commented:
@jacksonliam Since the memory layout of RGB and libjpeg are compatible its possible to use memcpy instead of the current loop, which could be more efficient:

```diff
     } else if (vd->formatIn == V4L2_PIX_FMT_RGB24) {
         while(cinfo.next_scanline < vd->height) {
-            int x;
             unsigned char *ptr = line_buffer;
-
-            for(x = 0; x < vd->width; x++) {
-                *(ptr++) = yuyv[0];
-                *(ptr++) = yuyv[1];
-                *(ptr++) = yuyv[2];
-                yuyv += 3;
-            }
-
+            int w = vd->width*3;
+            memcpy(line_buffer,yuyv,w);
+            yuyv += w;
             row_pointer[0] = line_buffer;
             jpeg_write_scanlines(&cinfo, row_pointer, 1);
         }
```

It didn't make a noticeable difference though, I will have to use a profiler.

GilesBathgate commented:
Actually there is no need for the memcpy at all:
```diff
         while(cinfo.next_scanline < vd->height) {
-            int x;
-            unsigned char *ptr = line_buffer;
-
-            for(x = 0; x < vd->width; x++) {
-                *(ptr++) = yuyv[0];
-                *(ptr++) = yuyv[1];
-                *(ptr++) = yuyv[2];
-                yuyv += 3;
-            }
-
-            row_pointer[0] = line_buffer;
+            int w = vd->width*3;
+            row_pointer[0] = yuyv;
             jpeg_write_scanlines(&cinfo, row_pointer, 1);
+            yuyv += w;
         }
```

jacksonliam commented:
Can possibly just take out the outer while loop and do 

Agreed that would work, probably slightly lower CPU usage but it's going to be minimal compared to the actual JPEG compression.

Armageddit commented:
![Uploading image.png…]()

Armageddit commented:
![Unbenan0nt](https://user-images.githubusercontent.com/49964721/80796501-a145f980-8b9f-11ea-9f94-f007562048d3.png)
